pandas/core/arrays/base.py

```diff
@@ -653,15 +660,46 @@ def copy(self, deep=False):
         raise AbstractMethodError(self)
 
     # ------------------------------------------------------------------------
-    # Block-related methods
+    # Printing
     # ------------------------------------------------------------------------
+    def __repr__(self):
+        from pandas.io.formats.printing import format_object_summary
+
+        template = (
+            '<{class_name}>\n'
+            '{data}\n'
+            'Length: {length}, dtype: {dtype}'
+        )
+        # the short repr has no trailing newline, while the truncated
+        # repr does. So we include a newline in our template, and strip
+        # any trailing newlines from format_object_summary
+        data = format_object_summary(self, self._formatter, name=False,
+                                     trailing_comma=False).rstrip('\n')
+        name = self.__class__.__name__
+        return template.format(class_name=name, data=data,
+                               length=len(self),
+                               dtype=self.dtype)
+
+    @property
+    def _formatter(self):
```

Comment:
Wouldn't it be possible to combine this with `_formatting_values` ?  

Or that makes it hard to use the  `format_object_summary` ?

Reply:
I gave up on sharing these. I think the order of operations is swapped between `format_object_summary` and `_formatting_values` (via internals).

IIUC, `format_object_summary` takes a whole array, truncates, then formats (need to get the length). But internals will truncate *before* calling `formatting_values`.

Reply:
Hmm, I understand that that's how our internals currently work, it's still a bit unfortunate to clutter our EA interface with it though ...

Would a not too big hack be a check inside `format_object_summary` for ExtensionArray, and in that case swap `obj` with `obj[selection]._formatting_values()` (where `selection` can just be the maximum values needed to select)

Reply:
@jorisvandenbossche 

> Hmm, I understand that that's how our internals currently work, it's still a bit unfortunate to clutter our EA interface with it though ...

its not unfortunate, rather its fortunate that we have a well test interface already. The EA interface IS lacking by-definition if you have to repeat code that we already have.


> would a not too big hack be a check inside format_object_summary for ExtensionArray, and in that case swap obj with obj[selection]._formatting_values() (where selection can just be the maximum values needed to select)

really try to avoid 'hacks'

Reply:
I think you misunderstood what Joris was saying. His point was it's unfortunate to have both `_formatting_values` and this new repr.

Reply:
> its not unfortunate, rather its fortunate that we have a well test interface already. The EA interface IS lacking by-definition if you have to repeat code that we already have.

Sorry Jeff, I don't understand your comment.

Reply:
> His point was it's unfortunate to have both _formatting_values and this new repr.

Small clarification: having both `_formatting_values` and `_formatter`. Having a default `__repr__` is great.

Reply:
And the reason that we in general should always be conservative in adding new methods here, is that this is all public interface (even if is a private attribute, it's part of the public EA interface)

Reply:
my comment means. that it is unfortunate that this code was repeated in the first place. The entire point of EA is to use our existing code that we *already* have as much as possible. I have complained about this before and will keep focusing on this here. Avoiding repeating the many may mistakes made in the repr here are paramount. Having repeated code is a distaster.

Reply:
@jorisvandenbossche how about we remove `_formatting_values` from the public interface, in favor of `_formatter`? Then in `ExtensionBlock._formatting_values`, we can do 

```python
def formatting_values(self):
    formatter = self.values._formatter
    return np.array([formatter(x) for x in self.values], dtype=object)
```

Reply:
Just to be sure (can't check right now): `_formatting_values` is only used for printing? And not for eg conversion to string in general (as that could give a performance penalty if the EA has a more efficient way to convert to strings)

And in the formatting code, are there some options that could influence the formatting (just thinking out loud, eg the datetime format)

Reply:
`_formatting_values` is only use is printing.

The trouble I'm running into now is an inconsistency between printing in Series and Index. Using (e.g.) datetimes, on master we quote values in the index, but not in the series.

```python
In [26]: pd.Series(pd.DatetimeIndex(['2001', '2002']))
Out[26]:
0   2001-01-01
1   2002-01-01
dtype: datetime64[ns]

In [27]: pd.DatetimeIndex(['2001', '2002'])
Out[27]: DatetimeIndex(['2001-01-01', '2002-01-01'], dtype='datetime64[ns]', freq=None)

In [28]: pd.Series(arr)
Out[28]:
0    '2000-01-01'
1    '2001-01-01'
dtype: period[D]

In [29]: pd.Index(arr)
Out[29]: PeriodIndex(['2000-01-01', '2001-01-01'], dtype='period[D]', freq='D')
```

Do we want to keep it like that? If so, `_formatter` would be a function that takes the "context" we're printing in?

Reply:
> Do we want to keep it like that? If so, _formatter would be a function that takes the "context" we're printing in (by "context", I mean whether we're inside a series / dataframe)?

I think I would personally be fine with dropping the quotes on the array repr as well, but for the Index repr it would of course be a bigger change ...